drivers/sensor/bq274xx/bq274xx.c

```diff
@@ -636,26 +637,156 @@ static int bq274xx_gauge_configure(const struct device *dev)
 		return -EIO;
 	}
 
+#ifdef CONFIG_PM_DEVICE
+	bq274xx->pm_state = PM_DEVICE_STATE_ACTIVE;
+#endif
+
+	return 0;
+}
+
+#ifdef CONFIG_PM_DEVICE
+static int bq274xx_enter_shutdown_mode(struct bq274xx_data *data)
+{
+	int status = 0;
+
+	status = bq274xx_control_reg_write(data, BQ274XX_UNSEAL_KEY);
+	if (status < 0) {
+		LOG_ERR("Unable to unseal the battery");
+		return -EIO;
+	}
+
+	status = bq274xx_control_reg_write(data, BQ274XX_UNSEAL_KEY);
+	if (status < 0) {
+		LOG_ERR("Unable to unseal the battery");
+		return -EIO;
+	}
+
+	status = bq274xx_control_reg_write(data,
+					   BQ274XX_CONTROL_SHUTDOWN_ENABLE);
+	if (status < 0) {
+		LOG_ERR("Unable to enable shutdown mode");
+		return -EIO;
+	}
+
+	status = bq274xx_control_reg_write(data, BQ274XX_CONTROL_SHUTDOWN);
+	if (status < 0) {
+		LOG_ERR("Unable to enter shutdown mode");
+		return -EIO;
+	}
+
+	status = bq274xx_control_reg_write(data, BQ274XX_CONTROL_SEALED);
+	if (status < 0) {
+		LOG_ERR("Failed to seal the gauge");
+		return -EIO;
+	}
+
+	data->pm_state = PM_DEVICE_STATE_OFF;
+
+	return 0;
+}
+
+static int bq274xx_exit_shutdown_mode(const struct device *dev)
+{
+	const struct bq274xx_config *const config = dev->config;
+	int status = 0;
+
+	if (!device_is_ready(config->gpio_pm.port)) {
+		LOG_ERR("GPIO device pointer is not ready to be used");
+		return -EINVAL;
+	}
+
+	status = gpio_pin_configure_dt(&config->gpio_pm,
+			   GPIO_OUTPUT | GPIO_OPEN_DRAIN);
+	if (status < 0) {
+		LOG_ERR("Unable to configure interrupt pin to output and open drain");
+		return status;
+	}
+
+	status = gpio_pin_set(config->gpio_pm.port, config->gpio_pm.pin, 0);
+	if (status < 0) {
+		LOG_ERR("Unable to set interrupt pin to low");
+		return status;
+	}
+
+	k_msleep(1);
+
+	status = gpio_pin_configure_dt(&config->gpio_pm,
+			   GPIO_INPUT);
+	if (status < 0) {
+		LOG_ERR("Unable to configure interrupt pin to input");
+		return status;
+	}
+
+	k_msleep(100);
+
+	status = bq274xx_gauge_configure(dev);
+	if (status < 0) {
+		LOG_ERR("Unable to configure bq274xx gauge");
+		return status;
+	}
+
 	return 0;
 }
 
+static int bq274xx_pm_control(const struct device *dev, uint32_t ctrl_command,
+				  void *context, pm_device_cb cb, void *arg)
+{
+	int ret = 0;
+	struct bq274xx_data *data = (struct bq274xx_data *)dev->data;
+
+	switch (ctrl_command) {
+	case PM_DEVICE_STATE_SET:
+		if (*((uint32_t *)context) == PM_DEVICE_STATE_OFF) {
+			ret = bq274xx_enter_shutdown_mode(data);
+			if (ret < 0) {
+				LOG_ERR("Unable to enter off state");
+			}
+		} else {
+			ret = bq274xx_exit_shutdown_mode(dev);
```

Comment:
States should be checked explicitly, `else` will catch everything, including the `suspend` case (leading to inconsistent behavior).

Reply:
another question: is the device being turned off or put into a low power mode?

Reply:
It's being turned off.

Reply:
I've added explicit `else if` for `PM_DEVICE_STATE_ACTIVE` state and I've added the `else` condition which prints out the log that the not implemented state is trying to be set.